File: app/src/main/java/com/example/mvvmdemo/MainActivity.java
package com.example.mvvmdemo;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import android.os.Bundle;
import android.view.View;

import com.example.mvvmdemo.adapters.RecyclerBindingDataAdapter;
import com.example.mvvmdemo.databinding.ActivityMainBinding;
import com.example.mvvmdemo.models.NicePlaces;
import com.example.mvvmdemo.viewmodels.MainActivityViewModel;

import java.util.List;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    ActivityMainBinding mainBinding;
    MainActivityViewModel mainActivityViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        mainActivityViewModel = new ViewModelProvider(this).get(MainActivityViewModel.class);
        mainActivityViewModel.init();

        RecyclerBindingDataAdapter recyclerBindingDataAdapter = new RecyclerBindingDataAdapter(this, mainActivityViewModel.getNicePlaces().getValue());
        mainBinding.setNicePlaceAdapter(recyclerBindingDataAdapter);
        mainBinding.setOnClickListener(this);

        mainActivityViewModel.getNicePlaces().observe(this, new Observer<List<NicePlaces>>() {
            @Override
            public void onChanged(List<NicePlaces> nicePlacesList) {
                recyclerBindingDataAdapter.notifyDataSetChanged();
            }
        });

        mainActivityViewModel.isLoading().observe(this, new Observer<Boolean>() {
            @Override
            public void onChanged(Boolean aBoolean) {
                if(aBoolean){
                    showProgressBar();
                }else{
                    hideProgressBar();
                }
                mainBinding.placeRecyclerView.smoothScrollToPosition(mainActivityViewModel.getNicePlaces().getValue().size() - 1);
            }
        });
    }

    private void showProgressBar(){
        mainBinding.progressBar.setVisibility(View.VISIBLE);
    }

    private void hideProgressBar(){
        mainBinding.progressBar.setVisibility(View.GONE);
    }

    @Override
    public void onClick(View v) {

        if(v == mainBinding.fab){
            mainActivityViewModel.addNewValue(new NicePlaces("Jammu", "http://d2c018txhbneoo.cloudfront.net/wakau/videoThumbnail/6188-1616173281-1625460864.jpg"));
        }
    }
}

File: app/src/main/java/com/example/mvvmdemo/activity/MainActivity.kt
package com.example.mvvmdemo.activity

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.example.mvvmdemo.R
import com.example.mvvmdemo.adapters.MainAdapter
import com.example.mvvmdemo.databinding.ActivityMainBinding
import com.example.mvvmdemo.models.AtcivityListModel
import com.example.mvvmdemo.viewmodels.MainActivityViewModel

class MainActivity : AppCompatActivity(), View.OnClickListener {
    var mainBinding: ActivityMainBinding? = null
    var mainActivityViewModel: MainActivityViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        mainActivityViewModel = ViewModelProvider(this).get(MainActivityViewModel::class.java)
        mainActivityViewModel!!.init()

        val recyclerBindingDataAdapter =
            MainAdapter(this, mainActivityViewModel!!.atcivityListModel!!.value as ArrayList)
        mainBinding?.let { it.nicePlaceAdapter = recyclerBindingDataAdapter }
        mainBinding?.let { it.onClickListener = this }

        mainActivityViewModel!!.atcivityListModel!!.observe(this,
            { recyclerBindingDataAdapter.notifyDataSetChanged() })

        mainActivityViewModel!!.isLoading.observe(this) { aBoolean ->
            if (aBoolean) {
                showProgressBar()
            } else {
                hideProgressBar()
            }
            mainBinding?.let { it.placeRecyclerView.smoothScrollToPosition(mainActivityViewModel!!.atcivityListModel!!.value!!.size - 1) }
        }
    }

    private fun showProgressBar() {
        mainBinding!!.progressBar.visibility = View.VISIBLE
    }

    private fun hideProgressBar() {
        mainBinding!!.progressBar.visibility = View.GONE
    }

    override fun onClick(v: View) {
        if (v === mainBinding!!.fab) {

        }
    }
}